File: 2024/python/days/day_3.py
from pathlib import Path

from utils.misc import timer
from utils.read import read_input
import re

example = """xmul(2,4)%&mul[3,7]!@^do_not_mul(5,5)+mul(32,64]then(mul(11,8)mul(8,5))"""
example_2 = """xmul(2,4)&mul[3,7]!^don't()_mul(5,5)+mul(32,64](mul(11,8)undo()?mul(8,5))"""

pattern = re.compile(r'mul\((\d{0,3}),(\d{0,3})\)')

def part1(input):
    answer = sum([int(a) * int(b) for a, b in re.findall(pattern, input)])
    print(f'Pt1::ans: {answer}')

def part2(input):
    i = 0
    answer = 0
    on = True
    while i < len(input):
        if input[i:].startswith('do()'):
            on = True
        elif input[i:].startswith("don't()"):
            on = False
        elif on:
            match = pattern.match(input[i:])
            if match:
                answer += int(match.group(1)) * int(match.group(2))
        i += 1
    print(f'Pt2::ans: {answer}')

def part2_fast(input):
    pos = 0
    on = True
    input_len = len(input)
    answer = 0
    while pos < input_len:
        if input[pos:pos +4] == "do()":
            on = True
            pos += 4
        elif input[pos:pos + 7] == "don't()":
            on = False
            pos += 7
        elif on:
            if input[pos:].startswith("mul("):
                pos += 4
                next_comma_idx = input[pos:].find(',')
                next_paren_idx = input[pos:].find(')')
                mul_len = next_paren_idx
                if next_comma_idx != -1 and next_paren_idx != -1 and next_comma_idx >= 1 and next_paren_idx > next_comma_idx and 3 <= mul_len and mul_len <= 7:
                    first = input[pos:pos + next_comma_idx]
                    second = input[pos + next_comma_idx + 1: pos + next_paren_idx]
                    if first.isdigit() and second.isdigit():
                        answer += int(first) * int(second)
                        pos += mul_len
            else:
                pos += 1
        else:
            pos += 1
    print(f'Pt2::ans: {answer}')

def run():
    day = Path(__file__).name.split('.')[0].split('_')[-1]
    input = read_input(day)
    with timer():
        part1(example)

    with timer():
        part1(input)

    with timer():
        part2(example_2)

    with timer():
        part2(input)

    with timer():
        part2_fast(input)

if __name__ == "__main__":
    run()

# Pt1::ans: 161
# Elapsed time: 17 microseconds
# Pt1::ans: 170068701
# Elapsed time: 231 microseconds
# Pt2::ans: 48
# Elapsed time: 24 microseconds
# Pt2::ans: 78683433
# Elapsed time: 12 ms

File: 2024/cpp23/days/day_3.cpp
#include <string>
#include <string_view>
#include <charconv>

#include <utils/misc.hpp>

std::string example_2(R"(xmul(2,4)&mul[3,7]!^don't()_mul(5,5)+mul(32,64](mul(11,8)undo()?mul(8,5)))");

bool can_parse_to_int(std::string_view sv) {
    int result;
    auto [ptr, ec] = std::from_chars(sv.data(), sv.data() + sv.size(), result);
    return ec == std::errc() && ptr == sv.data() + sv.size();
}

long to_long(std::string_view sv) {
    return std::strtol(sv.data(), nullptr, 10);
}

void part_ii(const std::string &input) {
    size_t pos = 0;
    bool on = true;
    std::string_view view(input);
    auto len = input.length();
    long ans = 0;
    while (pos < len) {
        if (view.substr(pos, 4) == "do()") {
            on = true;
            pos += 4;
        } else if (view.substr(pos, 7) == "don't()") {
            on = false;
            pos += 7;
        } else if (on) {
            if (view.substr(pos).starts_with("mul(")) {
                pos += 4;
                auto next_comma_idx = view.substr(pos).find(',');
                auto next_paren_idx = view.substr(pos).find(')');
                auto mul_len = next_paren_idx;
                if (next_comma_idx != std::string_view::npos && next_paren_idx != std::string_view::npos &&
                    next_comma_idx >= 1 &&
                    next_paren_idx > next_comma_idx && 3 <= mul_len && mul_len <= 7) {
                    auto first = view.substr(pos, next_comma_idx);
                    auto second = view.substr(pos + next_comma_idx + 1, next_paren_idx - next_comma_idx - 1);
                    if (can_parse_to_int(first) && can_parse_to_int(second)) {
                        ans += to_long(first) * to_long(second);
                        pos += mul_len;
                    }
                }
            } else {
                ++pos;
            }
        } else {
            ++pos;
        }
    }
    std::println("Part ii: {}", ans);
}

int main() {
    auto input = Utils::read_input(3);

    Utils::with_timer("Part ii (example)", []() {
        part_ii(example_2);
    });

    Utils::with_timer("Part ii (input)", [&input]() {
        part_ii(input);
    });
}